### ISSUE TITLE
Iceberg does not work with other types of data lakes

### ISSUE THREAD

**Issue by BsoBird** [apache/iceberg#8624]
### Apache Iceberg version

1.3.1 (latest release)

### Query engine

Spark

### Please describe the bug 🐞

I have found that when I configure multiple Catalogs in the same SparkSession, in some cases Iceberg does not work.
Example:
```
---iceberg catalog
spark.sql.catalog.datacenter	org.apache.iceberg.spark.SparkCatalog
spark.sql.catalog.datacenter.type	hadoop
spark.sql.catalog.datacenter.warehouse	/iceberg-catalog/warehouse

--paimon catalog
spark.sql.catalog.paimon	org.apache.paimon.spark.SparkCatalog
spark.sql.catalog.paimon.warehouse	hdfs:///paimon/warehouse
spark.jars	    /data/kyuubi/spark_aux_lib/paimon-spark-3.3-0.6-20230922.002014-17.jar
```
When I execute the MERGE INTO statement under ICEBERG CATALOG, the error message is as follows:
```
org.apache.spark.SparkException: The Spark SQL phase planning failed with an internal error. Please, fill a bug report in, and provide the full stack trace.
	at org.apache.spark.sql.execution.QueryExecution$.toInternalError(QueryExecution.scala:500)
	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:512)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$1(QueryExecution.scala:185)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.execution.QueryExecution.executePhase(QueryExecution.scala:184)
	at org.apache.spark.sql.execution.QueryExecution.sparkPlan$lzycompute(QueryExecution.scala:145)
	at org.apache.spark.sql.execution.QueryExecution.sparkPlan(QueryExecution.scala:138)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executedPlan$1(QueryExecution.scala:158)
	at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$2(QueryExecution.scala:185)
	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:510)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$1(QueryExecution.scala:185)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.execution.QueryExecution.executePhase(QueryExecution.scala:184)
	at org.apache.spark.sql.execution.QueryExecution.executedPlan$lzycompute(QueryExecution.scala:158)
	at org.apache.spark.sql.execution.QueryExecution.executedPlan(QueryExecution.scala:151)
	at org.apache.spark.sql.execution.QueryExecution.simpleString(QueryExecution.scala:204)
	at org.apache.spark.sql.execution.QueryExecution.org$apache$spark$sql$execution$QueryExecution$$explainString(QueryExecution.scala:249)
	at org.apache.spark.sql.execution.QueryExecution.explainString(QueryExecution.scala:218)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:103)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:169)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:95)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:98)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:94)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:584)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:176)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:584)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:267)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:263)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:560)
	at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:94)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted$lzycompute(QueryExecution.scala:81)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted(QueryExecution.scala:79)
	at org.apache.spark.sql.Dataset.<init>(Dataset.scala:220)
	at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:100)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:97)
	at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:622)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:617)
	at org.apache.kyuubi.engine.spark.operation.ExecuteStatement.$anonfun$executeStatement$1(ExecuteStatement.scala:83)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.kyuubi.engine.spark.operation.SparkOperation.$anonfun$withLocalProperties$1(SparkOperation.scala:155)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:169)
	at org.apache.kyuubi.engine.spark.operation.SparkOperation.withLocalProperties(SparkOperation.scala:139)
	at org.apache.kyuubi.engine.spark.operation.ExecuteStatement.executeStatement(ExecuteStatement.scala:78)
	at org.apache.kyuubi.engine.spark.operation.ExecuteStatement$$anon$1.run(ExecuteStatement.scala:100)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.AssertionError: assertion failed: No plan for ReplaceIcebergData RelationV2[data_from#13985, partner#13986, plat_code#13987, uni_shop_id#13988, category_id#13989, parent_category_id#13990, category_name#13991, root_cid#13992, is_leaf#13993, tenant#13994, last_sync#13995] datacenter.dwd.b_std_category, IcebergWrite(table=datacenter.dwd.b_std_category, format=ORC)
+- Sort [icebergbuckettransform(64, uni_shop_id#14019) ASC NULLS FIRST], false
   +- RepartitionByExpression [icebergbuckettransform(64, uni_shop_id#14019)], 12288
      +- Project [data_from#14016, partner#14017, plat_code#14018, uni_shop_id#14019, category_id#14020, parent_category_id#14021, category_name#14022, root_cid#14023, is_leaf#14024, tenant#14025, last_sync#14026]
         +- MergeRows[data_from#14016, partner#14017, plat_code#14018, uni_shop_id#14019, category_id#14020, parent_category_id#14021, category_name#14022, root_cid#14023, is_leaf#14024, tenant#14025, last_sync#14026, _file#14027]
            +- Join FullOuter, ((uni_shop_id#13979 = uni_shop_id#13988) AND (category_id#13980 = category_id#13989)), leftHint=(strategy=no_broadcast_hash)
               :- NoStatsUnaryNode
               :  +- Project [data_from#13985, partner#13986, plat_code#13987, uni_shop_id#13988, category_id#13989, parent_category_id#13990, category_name#13991, root_cid#13992, is_leaf#13993, tenant#13994, last_sync#13995, _file#14010, true AS __row_from_target#14013, monotonically_increasing_id() AS __row_id#14014L]
               :     +- Filter dynamicpruning#14075 [_file#14010]
               :        :  +- Project [_file#14074]
               :        :     +- Join LeftSemi, ((uni_shop_id#13979 = uni_shop_id#14066) AND (category_id#13980 = category_id#14067))
               :        :        :- Filter isnotnull(category_id#14067)
               :        :        :  +- RelationV2[uni_shop_id#14066, category_id#14067, _file#14074] datacenter.dwd.b_std_category
               :        :        +- Project [uni_shop_id#13979, category_id#13980]
               :        :           +- Filter ((rank#13984 = 1) AND (isnotnull(uni_shop_id#13979) AND isnotnull(category_id#13980)))
               :        :              +- Window [row_number() windowspecdefinition(shop_id#14000, cid#14001, modified#14004 DESC NULLS LAST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS rank#13984], [shop_id#14000, cid#14001], [modified#14004 DESC NULLS LAST]
               :        :                 +- Project [shop_id#14000 AS uni_shop_id#13979, cid#14001 AS category_id#13980, shop_id#14000, cid#14001, modified#14004]
               :        :                    +- HiveTableRelation [`dw_base_temp`.`category_analyse_result`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, Data Cols: [data_from#13997, partner#13998, plat_code#13999, shop_id#14000, cid#14001, parent_cid#14002, nam..., Partition Cols: []]
               :        +- RelationV2[data_from#13985, partner#13986, plat_code#13987, uni_shop_id#13988, category_id#13989, parent_category_id#13990, category_name#13991, root_cid#13992, is_leaf#13993, tenant#13994, last_sync#13995, _file#14010] datacenter.dwd.b_std_category
               +- Project [data_from#14028, partner#14029, plat_code#14030, uni_shop_id#13979, category_id#13980, parent_category_id#13981, category_name#13982, root_cid#14036, is_leaf#14037, tenant#14038, last_sync#13983, true AS __row_from_source#14015]
                  +- Filter (rank#13984 = 1)
                     +- Window [row_number() windowspecdefinition(shop_id#14031, cid#14032, modified#14035 DESC NULLS LAST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS rank#13984], [shop_id#14031, cid#14032], [modified#14035 DESC NULLS LAST]
                        +- Project [data_from#14028, partner#14029, plat_code#14030, shop_id#14031 AS uni_shop_id#13979, cid#14032 AS category_id#13980, parent_cid#14033 AS parent_category_id#13981, name#14034 AS category_name#13982, root_cid#14036, is_leaf#14037, tenant#14038, modified#14035 AS last_sync#13983, shop_id#14031, cid#14032, modified#14035]
                           +- HiveTableRelation [`dw_base_temp`.`category_analyse_result`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, Data Cols: [data_from#14028, partner#14029, plat_code#14030, shop_id#14031, cid#14032, parent_cid#14033, nam..., Partition Cols: []]

	at scala.Predef$.assert(Predef.scala:223)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.plan(QueryPlanner.scala:93)
	at org.apache.spark.sql.execution.SparkStrategies.plan(SparkStrategies.scala:69)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.$anonfun$plan$3(QueryPlanner.scala:78)
	at scala.collection.TraversableOnce$folder$1.apply(TraversableOnce.scala:196)
	at scala.collection.TraversableOnce$folder$1.apply(TraversableOnce.scala:194)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:199)
	at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:192)
	at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1431)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.$anonfun$plan$2(QueryPlanner.scala:75)
	at scala.collection.Iterator$$anon$11.nextCur(Iterator.scala:486)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:492)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.plan(QueryPlanner.scala:93)
	at org.apache.spark.sql.execution.SparkStrategies.plan(SparkStrategies.scala:69)
	at org.apache.spark.sql.execution.QueryExecution$.createSparkPlan(QueryExecution.scala:459)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$sparkPlan$1(QueryExecution.scala:145)
	at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$2(QueryExecution.scala:185)
	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:510)
	... 55 more
```

**Comment by BsoBird** [comment]
If I want to use multiple data lakes in the same session, is there a better way?

**Comment by RussellSpitzer** [comment]
Looks like the actual issue here is that the iceberg spark extensions are unable to work on a specific plan here. Someone would have to debug why that is, but in the meantime you could just not use the extensions.

**Comment by amogh-jahagirdar** [comment]
If you could share the actual SQL + details of the underlying table (partitioning/sorting) that would be helpful for a consistent repro and being able to debug. From the logical plan, the part that sticks out to me is all the plan nodes on top of MergeRows. I would've expected ReplaceIcebergData on top of the MergeRows, but maybe I'm missing something.

**Comment by BsoBird** [comment]
I can't run any MERGE INTO commands. This exception has nothing to do with table size or structure.

**Comment by BsoBird** [comment]
@amogh-jahagirdar 
I can't run any MERGE INTO commands. This exception has nothing to do with table size or structure.

**Comment by BsoBird** [comment]
SQL:
```
MERGE INTO datacenter.dwd.b_std_category t using(
select 
   data_from,
   partner,
   plat_code,
   uni_shop_id,
   category_id,
   parent_category_id,
   category_name,
   root_cid,
   is_leaf,
   tenant,
   last_sync
   from
    (
          select
             data_from,
             partner,
             plat_code,
             shop_id as uni_shop_id,
             cid as category_id,
             parent_cid as parent_category_id,
             name as category_name,
             root_cid,
             is_leaf,
             tenant,
             modified as last_sync,
             row_number() over(partition by shop_id,cid order by modified desc) rank
          from spark_catalog.dw_base_temp.category_analyse_result
    ) small 
    where rank=1
)s
on s.uni_shop_id = t.uni_shop_id and s.category_id = t.category_id
WHEN MATCHED AND s.last_sync>=t.last_sync
THEN UPDATE SET
t.data_from            =  s.data_from           ,
t.partner              =  s.partner             ,
t.plat_code            =  s.plat_code           ,
t.uni_shop_id          =  s.uni_shop_id         ,
t.category_id          =  s.category_id         ,
t.parent_category_id   =  s.parent_category_id  ,
t.category_name        =  s.category_name       ,
t.root_cid             =  s.root_cid            ,
t.is_leaf              =  s.is_leaf             ,
t.tenant               =  s.tenant              ,
t.last_sync            =  s.last_sync           
WHEN NOT MATCHED THEN INSERT *
```

create sql:
```
CREATE TABLE IF NOT EXISTS datacenter.dwd.b_std_category (
  data_from int,
  partner string,
  plat_code string,
  uni_shop_id string,
  category_id string,
  parent_category_id string,
  category_name string,
  root_cid string,
  is_leaf int,
  tenant string,
  last_sync string
)
PARTITIONED BY (bucket(64,uni_shop_id))
TBLPROPERTIES ('read.orc.vectorization.enabled'='true','write.format.default'='orc','write.orc.bloom.filter.columns'='category_id,root_cid','write.orc.compression-codec'='zstd','write.metadata.previous-versions-max'='3','write.metadata.delete-after-commit.enabled'='true') 
STORED AS iceberg;
```

**Comment by BsoBird** [comment]
Paimon jar:
```
https://repository.apache.org/content/groups/snapshots/org/apache/paimon/paimon-spark-3.3/0.6-SNAPSHOT/paimon-spark-3.3-0.6-20230922.002014-17.jar
```

**Comment by amogh-jahagirdar** [comment]
> @amogh-jahagirdar I can't run any MERGE INTO commands. This exception has nothing to do with table size or structure.

Yes I understand doesn't have anything to do with size but I'm asking for the structure details so that we can reproduce hitting this particular plan more accurately. 

Or are you saying any MERGE INTO just with the paimon catalog configured fails with this error?

**Comment by BsoBird** [comment]
--Or are you saying any MERGE INTO just with the paimon catalog configured fails with this error?

**Comment by BsoBird** [comment]
Incidentally, when will version 1.4.0 be available? We would really like to use distributed planning.

**Comment by amogh-jahagirdar** [comment]
Thanks for confirming @BsoBird that gives us a wider range then to be able to repro this problem :)

> Incidentally, when will version 1.4.0 be available? We would really like to use distributed planning.

We expect a release candidate out for vote some time soon. You can track the pending items here https://github.com/apache/iceberg/milestone/35

**Comment by BsoBird** [comment]
@amogh-jahagirdar  
 Hi. Anything new on this subject?

**Comment by BsoBird** [comment]
@RussellSpitzer @amogh-jahagirdar   Hi. Anything new on this subject?

**Comment by BsoBird** [comment]
Paimon seems to have solved the problem, I updated to version 0.7 and the problem disappeared
